### PR TITLE
Switch to server_group for 18 autoscaling

### PR DESCRIPTION
**Pull request by vyzigold**
Latest updates to ceilometer and sg-core allow us to use server_group for autoscaling server grouping instead of the server_name_prefix. Unfortunately this functionality is only on ceilometer master, so we need to patch the openstackversions to use that. Because ceilometer-central-agent is running on the compute nodes, we need to redeploy the dataplane after patching the openstackversions.

I also found a few TODOs and other places that I thought I can enhance a little. I included that as well.

### CODE REVIEW

File: roles/telemetry_autoscaling/tasks/creating_stack.yml [hunk 1]
```diff
@@ -63,7 +63,7 @@
     # source ~/overcloudrc;
     {{ openstack_cmd }} metric list
   register: result
-  failed_when: result.rc >= 1
+  failed_when: '"ceilometer_cpu" not in result.stdout'
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/creating_stack.yml:66]: The original didn't do much checking. I think this condition is better as it does what's written in the task name.

File: roles/telemetry_autoscaling/tasks/creating_stack.yml [hunk 2]
```diff
@@ -82,7 +82,7 @@
     # source ~/overcloudrc;
     {{ openstack_cmd }} alarm list
   register: result
-  failed_when: result.rc >= 1
+  failed_when: '"cpu_alarm_high" not in result.stdout or "cpu_alarm_low" not in result.stdout'
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/creating_stack.yml:85]: The original didn't do much checking. I think this condition is better as it does what's written in the task name.

File: roles/telemetry_autoscaling/tasks/creating_stack.yml [hunk 3]
```diff
@@ -101,20 +101,19 @@
 - name: RHOSO-12653 Verify physical_resource_id match the alarm id for cpu_alarm_low
   ansible.builtin.shell: |
     # source ~/overcloudrc;
-    {{ openstack_cmd }} alarm list |grep -i cpu_alarm_low | awk '{print $2}'
+    {{ openstack_cmd }} alarm list |grep -i cpu_alarm_low | grep {{ stack_name }} | awk '{print $2}'
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/creating_stack.yml:104]: I added a grep for the stack_name. This is just to allow to run these tests when there is already an autoscaling stack with a different name present in the cluster.

File: roles/telemetry_autoscaling/tasks/creating_stack.yml [hunk 4]
```diff
@@ -101,20 +101,19 @@
...
   register: alarm_id_low
   failed_when:
     - physical_resource_id_low.stdout != alarm_id_low.stdout
 
 - name: RHOSO-12654 Verify physical_resource_id match the alarm id for cpu_alarm_high
   ansible.builtin.shell: |
     # source ~/overcloudrc;
-    {{ openstack_cmd }} alarm list |grep -i cpu_alarm_high | awk '{print $2}'
+    {{ openstack_cmd }} alarm list |grep -i cpu_alarm_high | grep {{ stack_name }} | awk '{print $2}'
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/creating_stack.yml:112]: I added a grep for the stack_name. This is just to allow to run these tests when there is already an autoscaling stack with a different name present in the cluster.

File: roles/telemetry_autoscaling/tasks/post_teardown.yml [hunk 5]
```diff
@@ -14,10 +14,9 @@
 
 - name: Wait until the stack is deleted
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/post_teardown.yml:15]: I don't think this ever worked correctly. In a situation when this should pass, the original output.stdout_lines would be an empty list, not 0. I checked a recent CI run and this actually failed. I added | wc -l so the check is now similar to a few other ones from this role.

File: roles/telemetry_autoscaling/tasks/test_autoscaling.yml [hunk 6]
```diff
@@ -62,6 +62,7 @@
     # source ~/overcloudrc;
     {{ openstack_cmd }} alarm list -c state -c name -f value| \
         grep -i "cpu_alarm_high" | \
+        grep -i "{{ stack_name }}" | \
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/test_autoscaling.yml:65]: I added a grep for the stack_name. This is just to allow to run these tests when there is already an autoscaling stack with a different name present in the cluster.

File: roles/telemetry_autoscaling/tasks/test_autoscaling.yml [hunk 7]
```diff
@@ -92,20 +93,18 @@
     # source ~/overcloudrc;
     {{ openstack_cmd }} alarm list -c state -c name -f value| \
         grep -i "cpu_alarm_low" | \
+        grep -i "{{ stack_name }}" | \
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/test_autoscaling.yml:96]: I added a grep for the stack_name. This is just to allow to run these tests when there is already an autoscaling stack with a different name present in the cluster.

File: roles/telemetry_autoscaling/tasks/test_autoscaling.yml [hunk 8]
```diff
@@ -92,20 +93,18 @@
...
         awk '{print $2}'
   retries: 100
   delay: 5
   register: result
   until: result.stdout == "alarm"
 
-  # TODO: the metering.server group metadata was used for gnocchi alarm
-  # selection.
-  # prom uses the instance name, so the metadata MIGHT be removed, and a new
-  # check for whether the scaling group has scaled down may be needed.
```
Comment (vyzigold) [inline roles/telemetry_autoscaling/tasks/test_autoscaling.yml:104]: This is now actually the prefered way.

File: ci/run_autoscaling_osp18.yml [hunk 9]
```diff
@@ -62,3 +85,14 @@
               oc patch  openstackversions controlplane  --type json -p='[{"op": "replace", "path": "/spec/customContainerImages", "value": {} }]'
           when: "{{ patch_openstackversions | bool }}"
 
+        - name: Redeploy the dataplane
```
Comment (elfiesmelfie) [inline ci/run_autoscaling_osp18.yml:88]: Why is the dataplane being redeployed twice?

File: roles/telemetry_autoscaling/tasks/post_teardown.yml (hunk 5 above)
Comment (elfiesmelfie) [reply roles/telemetry_autoscaling/tasks/post_teardown.yml:15]: This should never have worked. I suspect that the ``ignore_errors: true`` is also not needed, and may have been the reason that this wasn't causing issues

File: ci/run_autoscaling_osp18.yml (hunk 9 above)
Comment (vyzigold) [reply ci/run_autoscaling_osp18.yml:88]: It's redeployed again to revert the changes done by the first redeployment. So after the first time we have master ceilometer containers. After the second time we have antelope containers again.

Comment (vyzigold) [reply ci/run_autoscaling_osp18.yml:88]: right

Comment (vyzigold) [reply ci/run_autoscaling_osp18.yml:88]: Dataplane is updated only as a reaction to a new OpenstackDataplaneDeployment CR. That's why there are tasks to copy the current osdpd into a file, delete the osdpd and apply it again from the file.

**Review by elfiesmelfie — APPROVED** [review]

**Comment by vyzigold** [comment]
recheck graphing test failure.

**Pull request merged**